src/Servers/Kestrel/shared/HPackHeaderWriter.cs

```diff
@@ -115,22 +127,23 @@ private static bool EncodeHeadersCore(DynamicHPackEncoder hpackEncoder, Http2Hea
                 out var headerLength))
             {
                 // If the header wasn't written, and no headers have been written, then the header is too large.
-                // Throw an error to avoid an infinite loop of attempting to write large header.
-                if (currentLength == 0 && throwIfNoneEncoded)
+                // Request for a larger buffer to write large header.
+                if (currentLength == 0 && canRequestLargerBuffer)
                 {
-                    throw new HPackEncodingException(SR.net_http_hpack_encode_failure);
+                    length = 0;
+                    return HeaderWriteResult.BufferTooSmall;
                 }
 
                 length = currentLength;
-                return false;
+                return HeaderWriteResult.MoreHeaders;
```

Comment:
what is the behaviour here if we *can't* request more (`canRequestLargerBuffer == false`) ? I haven't pulled the code locally, but it *looks* like we will return `MoreHeaders`, which will ... keep trying? is there any way this can result in an infinite loop of not making progress?

Reply:
This (that it cannot request more) should be only the case for the first HEADER frame which always has status response, so no loop. We could consider completely removing this though.

Reply:
There is a debug assert on the call site.

Reply:
The `canRequestLargerBuffer` could dropped if we apply a do-while loop on the initial header as well, but I felt that might sacrifice perf. I will do a measurement.

Reply:
But if I did that, remove `canRequestLargerBuffer` and implement a do-while loop in `WriteResponseHeadersUnsynchronized` to enlarge the header, the perf decreases. Using the same `Http2FrameWriterBenchmark` as in the description:

|               Method |     Mean |    Error |   StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |---------:|---------:|---------:|-------------:|-------:|------:|------:|----------:|
| WriteResponseHeaders | 80.23 ns | 1.206 ns | 1.128 ns | 12,463,819.6 | 0.0002 |     - |     - |      32 B |